server/src/main/java/org/opensearch/index/mapper/NumberFieldMapper.java

```diff
@@ -988,23 +988,34 @@ public Query rangeQuery(
                         --u;
                     }
                 }
-                if (isSearchable && hasDocValues) {
-                    Query query = IntPoint.newRangeQuery(field, l, u);
-                    Query dvQuery = SortedNumericDocValuesField.newSlowRangeQuery(field, l, u);
-                    query = new IndexOrDocValuesQuery(query, dvQuery);
-                    if (context.indexSortedOnField(field)) {
-                        query = new IndexSortSortedNumericDocValuesRangeQuery(field, l, u, query);
+                Query dvQuery = hasDocValues ? SortedNumericDocValuesField.newSlowRangeQuery(field, l, u) : null;
```

Comment:
Can we refactor these `rangeQuery` methods to cut down on the duplication?

I'm thinking that the type-specific information is:
1. What are the upper/lower limits for the type?
2. How do I parse values into the given type? (This is already captured in the `parse` method.)
3. How do I create a `PointRangeQuery` for the type?
4. How do I represent the type as a `long` for use in DV queries (both in `SortedNumericDocValuesField.newSlowRangeQuery` and `IndexSortSortedNumericDocValuesRangeQuery`)?
5. How do I pack values of the type into point value bytes? <-- In theory, this can be used to produce a default implementation for number 3.

Reply:
I would lean toward splitting this into two PRs.

The traversal optimization is one thing. Handling all the other types (by doing the above refactoring) is another.